### ISSUE TITLE
Rule proposal: empty-line-between-distinct-statements

### ISSUE THREAD

**Issue by gajus** [eslint/eslint#7655]
**Please describe what the rule should do:**

Rule enforces use of an empty line between distinct statements.

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// Good

window.adTrackingConfig = createAdTrackingConfig([
  'grapeshot'
]);

AdRunner.setup();

AdsTrackers.runTrackingPlugins();

const gsVar0 = 'gs_name';
const gsVar1 = 'gs_value';

window[gsVar0] = [
  'color',
  'size'
];
window[gsVar1] = [
  'red',
  '468x60'
];

const requiredDependencies = [
  'apple',
  'bannana'
];
const expectedTags = [
  {
    grapeshot: [
      'value1',
      'value2'
    ]
  }
];

await resolveDependencies(requiredDependencies);
await requestAndCompareTags({}, expectedTags, t.deepEqual.bind(t));
```

```js
// Bad

window.adTrackingConfig = createAdTrackingConfig([
  'grapeshot'
]);

// There should be an empty line after `AdRunner` invocation
// because `AdRunner` and `AdsTrackers` are distinct variables.
AdRunner.setup();
AdsTrackers.runTrackingPlugins();

const gsVar0 = 'gs_name';
const gsVar1 = 'gs_value';

window[gsVar0] = [
  'color',
  'size'
];

// There should be an empty line after window
// because window and const are distinct statements.
window[gsVar1] = [
  'red',
  '468x60'
];
const requiredDependencies = [
  'apple',
  'bannana'
];
const expectedTags = [
  {
    grapeshot: [
      'value1',
      'value2'
    ]
  }
];

await resolveDependencies(requiredDependencies);
await requestAndCompareTags({}, expectedTags, t.deepEqual.bind(t));
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It is a fairly common style requirement.

**Comment by gajus** [comment]
Rule name needs to be improved though. This is broader than just distinct statements.

**Comment by gajus** [comment]
Closing as this was deemed too opinionated.

https://gitter.im/eslint/eslint?at=58385c21b9016e42149a1d3f
